src/worker/ingestion/process-event.ts

```diff
@@ -343,26 +345,48 @@ export class EventsProcessor {
 
         await this.db.updatePerson(mergeInto, { created_at: firstSeen, properties: mergeInto.properties })
 
-        // merge the distinct_ids
+        // Merge the distinct IDs
         for (const otherPerson of peopleToMerge) {
-            const otherPersonDistinctIds: PersonDistinctId[] = (
+            let failedAttempts = 0
+            // Retrying merging up to `MAX_FAILED_PERSON_MERGE_ATTEMPTS` times, in case race conditions occur.
+            // An example is a distinct ID being aliased in another plugin server instance,
+            // between `moveDistinctId` and `deletePerson` being called here
+            // – in such a case a distinct ID may be assigned to the person in the database
+            // AFTER `otherPersonDistinctIds` was fetched, so this function is not aware of it and doesn't merge it.
+            // That then causeds `deletePerson` to fail, because of foreign key constraints –
+            // the dangling distinct ID added elsewhere prevents the person from being deleted!
+            // This is low-probability so likely won't occur on second retry of this block.
+            // In the rare case of the person changing VERY often however, it may happen even a few times,
+            // in which case we'll bail and rethrow the error.
+            while (true) {
+                const otherPersonDistinctIds: PersonDistinctId[] = (
+                    await this.db.postgresQuery(
+                        'SELECT * FROM posthog_persondistinctid WHERE person_id = $1 AND team_id = $2',
+                        [otherPerson.id, mergeInto.team_id],
+                        'otherPersonDistinctIds'
+                    )
+                ).rows
+                for (const personDistinctId of otherPersonDistinctIds) {
+                    await this.db.moveDistinctId(otherPerson, personDistinctId, mergeInto)
```

Comment:
I'm a bit wary of this. With retries now, `moveDistinctId` is called multiple times with the same `otherPerson`, `personDistinctId`. (more plausible when messages are sent to kafka, instead of directly operating on the db, since then the first postgres query returns the same rows).

This might lead to new race conditions where follow-up moves are overridden by this extra move. I'm not sure if this can actually happen, and it seems less likely than the existing race condition, but still wanted to raise it.

Reply:
If I understood what you mean correctly, this shouldn't be a problem, because `moveDistinctId` wouldn't be called multiple times with the same `personDistinctId`  – the retry's `otherPersonDistinctIds` won't contain already moved IDs, as Postgres is the source of truth in all these queries, and we only need to move the IDs not moved on previous iteration due to being added after previous iteration started. Did I miss something?

Reply:
Ah, I see, I misread the code. When we send kafka messages, we update Postgres *before* sending these messages. This makes sense, there shouldn't be a problem here then!